daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -657,6 +657,25 @@ object Transaction {
       consumedBy: NodeId,
   ) extends TransactionError
 
+  /** Signals that within the transaction we got to a point where
+    * two contracts with the same key were active.
+    *
+    * Note that speedy only detects duplicate key collisions
+    * if both contracts are used in the transaction (fetched/created/looked up/…).
+    * In particular, a standalone create will never produce a duplicate key
+    * error if the global contract with the duplicate key is not used in the same
+    * transaction.
+    *
+    * At the moment, Speedy also only if the two contracts are actually active at the same time.
+    * It makes no attempts to detect contract key inconsistencies, e.g.,
+    * a local create with key A, followed by an archive, followed by a global fetch
+    * with key A. This is just a more complex version of the standalone create where
+    * the global key is not fetched at all and should be caught by the ledger.
+    */
```

Comment:
Thanks for the ping @cocreature . Do you want to say that you the engine only checks internal consistency wrt contract keys (cf. https://docs.daml.com/concepts/ledger-model/ledger-integrity.html#internal-consistency) and this error signals an `InternallyInconsistentContractKey`?

Reply:
Not really sure what you mean by that exactly. The ledger model does not define a concept of internal key consistency. 

What I’m trying to say is that we detect if we observe two contracts being active with the same key at the same time. 

What we do not do atm is to make inferences about past activeness. In particular, for a global contract we know that even if we have a later fetch it must have been active at the beginning of the transaction as well, so the following transaction will not be rejected by speedy:
```
cid <- create c -- with key K
archive cid
fetch globalCid -- with key k
```
 We could detect those and while I don’t feel strongly about this, I’d be very open to agreeing that speedy should reject this as well but I’d prefer to keep it out of this PR since it’s somewhat orthogonal and this is large enough as it is.

Reply:
Ignore my comment wrt internal key consistency. The definition in https://docs.daml.com/concepts/ledger-model/ledger-integrity.html#internal-consistency, reads

> A transaction is *internally consistent* if it is internally consistent for all contracts and consistent for all keys.

So, it refers to the (global) consistency notion for keys, which matches with what you are aiming for. Your example
```
cid <- create c -- with key K 
archive cid 
fetch globalCid -- with key k
```
does though not satisfy /key consistency/ and should therefore be rejected by the engine as the resulting transaction is not internally consistent.

Reply:
Makes sense, as I said, I’m totally fine with rejecting this in speedy. I would just prefer to keep it out of this PR since it only makes things more confusing.

Reply:
Thanks, good to see this tightened up.